src/manager.rs

```diff
@@ -165,14 +165,69 @@ impl Manager {
         Ok(config)
     }
 
-    /// Spawns all units and targets in the config unto the given runtime.
+    /// Allows creating units and adding them to the manager.
+    ///
+    /// Because creating units that contain links requires some setup work,
+    /// this has to happen inside a closure. This closure should return a
+    /// list of units 
+    pub fn add_units<F, T>(
+        &mut self, runtime: &runtime::Handle, op: F
+    ) -> Result<T, Failed>
+    where
+        F: FnOnce(&mut UnitSet, &mut TargetSet) -> T
+    {
+        GATES.with(|gates| {
```

Comment:
`GATES` is actually used in the `FromStr` impl for `Link` which you are using inside `add_components` to create those links.

Reply:
Sorry, I must be being blind, I can't see any use of `FromStr` between this line which populates `GATES` and the `let gates = GATES.with(|gates| gates.replace(None)).unwrap()` line below which takes the value out of `GATES`, except perhaps the invocation of `op()` (but the example use of `op()` in `test.rs` that I looked at also doesn't use `FromStr` that I can see). Actually, I'm super confused now. This call to `gates.replace()` uses `self.units` but in `tests.rs` it does:
```
let mut manager = Manager::new();
let (u, mut t) = manager.add_units(...)
```
Nothing is done betwenen the call to `new()` which creates an empty `self.units` `HashMap` and the call to `add_units()` which then immediately uses `self.units.iter()` (which must be empty) in the call to `GATES.wtih(...)`...

Reply:
Sorry, `Link` isn’t using `FromStr` just plain old `From<_>` for anything string-like. Because these calls happen the links are created they have to happen inside the closure because they needs the thread local.

In the `test.rs` test, it’s this line:
```
let (t, tc) = Target::new("u");
```
The string `"u"` gets converted into a link.

Reply:
Ah I only looked at `Unit::new()`, not `Target::new()`. Thanks.

Reply:
What seems to missing is to add the newly created units (or rather their gate agents) to `self.units`. I guess we never noticed since there is no dynamic re-configuration yet.

Reply:
Ok. I suppose as long as there way to trigger `add_components()` on a running graph this is fine, because if  you try and do that you can't add a new target for an existing unit as the existing units Gate can't be linked to by `load_link()` (as `gates.entry(name).or_default()` triggers the `or_default()` branch, because `GATES` was populated with `value.clone().into()` on the `GateAgent` which causes an "empty" link to be constructed), and you can't replace an existing unit either. You can add a new target for an existing unit, that seems to work and receive updates sent via the original gate, and you can add a new unit and target pair which also seems to work.

Reply:
I don’t quite follow. `GATES` will have gate agents for all existing units so when you call `gates.entry()` you get the pre-populated value not a new default and can create a link from that?

Reply:
I was confused, ignore.